common/djangoapps/oauth_exchange/forms.py

```
@@ -0,0 +1,109 @@
+
+    def _require_oauth_field(self, field_name):
+        field_val = self.cleaned_data.get(field_name)
+        if not field_val:
+            raise OAuthValidationError(
+                {
+                    "error": "invalid_request",
+                    "error_description": "{} is required".format(field_name),
+                }
+            )
+        return field_val
+
+    def clean_provider(self):
+        return self._require_oauth_field("provider")
+
+    def clean_access_token(self):
+        return self._require_oauth_field("access_token")
+
+    def clean_client_id(self):
+        return self._require_oauth_field("client_id")
+
+    def clean(self):
+        if self._errors:
+            return {}
+
+        provider_name = self.cleaned_data["provider"]
+        provider_class = Registry.get_by_backend_name(provider_name)
+        if not (
+                provider_class and
+                issubclass(provider_class.BACKEND_CLASS, social_oauth.BaseOAuth2)
+        ):
+            raise OAuthValidationError(
+                {
+                    "error": "invalid_request",
+                    "error_description": "{} is not a supported provider".format(provider_name),
+                }
+            )
+
+        self.request.session[pipeline.AUTH_ENTRY_KEY] = pipeline.AUTH_ENTRY_API
+        strategy = social_utils.load_strategy(
+            request=self.request,
+            backend=provider_class.BACKEND_CLASS.name
+        )
+        backend = strategy.backend
+
+        client_id = self.cleaned_data["client_id"]
+        try:
+            client = Client.objects.get(client_id=client_id)
+        except Client.DoesNotExist:
+            raise OAuthValidationError(
+                {
+                    "error": "invalid_client",
+                    "error_description": "{} is not a valid client_id".format(client_id),
+                }
+            )
+        if client.client_type != provider.constants.PUBLIC:
```

Then in order to remain consistent, why not return the same error type `invalid_client` as `PublicPasswordGrantForm.clean`?